internal/controllers/cryostat_controller.go

```diff
@@ -659,6 +676,66 @@ func (r *CryostatReconciler) deleteConsoleLink(ctx context.Context, cr *operator
 	return nil
 }
 
+func (r *CryostatReconciler) addCorsAllowedOriginIfNotPresent(allowedOrigin string, cr *operatorv1beta1.Cryostat) error {
+	reqLogger := r.Log.WithValues("Request.Namespace", cr.Namespace, "Request.Name", cr.Name)
+	apiServer := &configv1.APIServer{}
+	err := r.Client.Get(context.Background(), types.NamespacedName{Name: resources.ApiServerName}, apiServer)
+	if err != nil {
+		reqLogger.Error(err, "Failed to get APIServer config")
+		return err
+	}
+
+	allowedOriginAsRegex := regexp.QuoteMeta(allowedOrigin)
+
+	for _, origin := range apiServer.Spec.AdditionalCORSAllowedOrigins {
+		if origin == allowedOriginAsRegex {
+			return nil
+		}
+	}
+
+	apiServer.Spec.AdditionalCORSAllowedOrigins = append(
+		apiServer.Spec.AdditionalCORSAllowedOrigins,
+		allowedOriginAsRegex,
+	)
+
+	err = r.Client.Update(context.Background(), apiServer)
```

Comment:
I noticed this `Update` call is also adding empty fields for `spec.ClientCA.Name`, `spec.encryption` and `spec.servingCerts`. The docs mention the APIServer will use default behaviour if these fields are "unset", which might be different than having an empty value.

I also tried manually editing the `APIServer` to remove the empty fields. Sometimes I get the below error, and sometimes I can remove the fields. Removing the fields manually has no effect on the pending cryostat pod.
```
jalaw@localhost ~/cryostat-operator (oauth-logout)> oc edit apiserver cluster
Error from server (InternalError): an error on the server ("") has prevented the request from succeeding (get apiservers.config.openshift.io cluster)
```
APIServer spec before running `make create_cryostat_cr`:
```
oc get apiserver cluster -o yaml
...
spec:
  audit:
    profile: Default
```

APIServer spec after running `make create_cryostat_cr`:
```
oc get apiserver cluster -o yaml
...
spec:
  additionalCORSAllowedOrigins:
  - https://cryostat-sample-default\.apps\.jalaw1\.lab\.upshift\.rdu2\.redhat\.com
  audit:
    profile: Default
  clientCA:
    name: ""
  encryption: {}
  servingCerts: {}
```

Reply:
I wonder if the empty spec fields are causing some of these errors below?

while running `make deploy`:
```
error: You must be logged in to the server (Unauthorized)
error: couldn't get available api versions from server: Unauthorized
make: *** [Makefile:175: check_cert_manager] Error 1

$ oc login
Error from server (InternalError): Internal error occurred: unexpected response: 400
$ oc login --loglevel=10
...
server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "Internal error occurred: unexpected response: 400",
  "reason": "InternalError",
  "details": {
    "causes": [
      {
        "message": "unexpected response: 400"
      }
    ]
  },
  "code": 500
}]
F0111 13:05:20.259012  149235 helpers.go:115] Error from server (InternalError): Internal error occurred: unexpected response: 400
...
```
```
oc get pods -w
NAME                                                    READY   STATUS    RESTARTS   AGE
cryostat-operator-controller-manager-78c899f994-qhtdm   1/1     Running   7          68m
cryostat-sample-86468b74d6-xcjj2                        0/3     Pending   0          47s
^C⏎                                                                                                                                                                                                                                                              jalaw@localhost ~/cryostat-operator (oauth-logout) [1]> oc logs -p cryostat-operator-controller-manager-78c899f994-qhtdm
2022-01-11T19:01:56.920Z	ERROR	controller-runtime.manager	Failed to get API Group-Resources	{"error": "Get \"https://172.30.0.1:443/api?timeout=32s\": dial tcp 172.30.0.1:443: connect: connection refused"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
sigs.k8s.io/controller-runtime/pkg/manager.New
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/manager/manager.go:317
main.main
	/workspace/internal/main.go:92
runtime.main
	/usr/local/go/src/runtime/proc.go:204
2022-01-11T19:01:56.920Z	ERROR	setup	unable to start manager	{"error": "Get \"https://172.30.0.1:443/api?timeout=32s\": dial tcp 172.30.0.1:443: connect: connection refused"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
main.main
	/workspace/internal/main.go:102
runtime.main
	/usr/local/go/src/runtime/proc.go:204
```

Reply:
FWIW, I tried this PR on a test cluster other than CRC, and it didn't cause any outages or errors.

Reply:
Thanks, all the errors disappeared after switching to cluster-bot.